File: lib/kmc/versioner.rb
module Kmc
  class Versioner
    class << self
      INIT_COMMIT_MESSAGE = "INIT: Initialize KMC"

      def init_repo(path)
        GitAdapter.init_repo(path)
        GitAdapter.commit_everything(path, INIT_COMMIT_MESSAGE)
      end

      def mark_preinstall(path, package)
        GitAdapter.commit_everything(path, pre_install_message(package))
      end

      def mark_postinstall(path, package)
        GitAdapter.commit_everything(path, post_install_message(package))
      end

      def installed_packages(path)
        commits = GitAdapter.list_commits(path)

        postinstalls = commits.select(&:post?).map(&:subject)
        uninstalls = commits.select(&:uninstall?).map(&:subject)

        # The packages that have actually been installed are those that have a
        # 'post-install' without any corresponding 'uninstall'. We can find
        # these packages by getting post-installs, and removing the first
        # instance of each package also found in uninstall.
        #
        # Note that we can't use Array#delete because that method will delete
        # *all* instances of a package, which won't work because it's possible
        # for a user to install, uninstall, then re-install a package.
        uninstalls.each do |package|
          postinstalls.delete_at(postinstalls.index(package))
        end

        postinstalls.map { |package_title| Package.find(package_title) }
      end

      # Has this package already been installed?
      def already_installed?(path, package)
        installed_packages(path).include?(package)
      end

      def uninstall_package(path, package)
        to_revert = GitAdapter.list_commits(path).find do |commit|
          commit.post? && commit.subject == package.title
        end

        GitAdapter.revert_commit(path, to_revert)
        Util.run_post_processors!(path)
        GitAdapter.commit_everything(path, uninstall_message(package))
      end

      private

      def pre_install_message(package)
        "PRE: #{package.title}"
      end

      def post_install_message(package)
        "POST: #{package.title}"
      end

      def uninstall_message(package)
        "UNINSTALL: #{package.title}"
      end
    end
  end
end

File: lib/versioner.js
// A thin interface above `GitAdapter`, this module provides a simple API for
// working with version control for KMC's purposes.
var Versioner = {}

var _ = require('lodash');

var GitAdapter = require('./git_adapter');

// Sets up a repo at `kspPath` by creating the git repo and committing everything
// into an initial commit.
Versioner.setUpRepo = function(kspPath) {
  return GitAdapter.initRepo(kspPath).then(function() {
    return GitAdapter.commitEverything(kspPath, Versioner.initMessage());
  });
}

// Marks the pre-install step for installing `package` into a KSP located at
// `kspPath`.
Versioner.markPreinstall = function(kspPath, package) {
  return GitAdapter.commitEverything(kspPath,
    Versioner.preinstallMessage(package));
}

// Marks the post-install step for installing `package` into a KSP located at
// `kspPath`.
Versioner.markPostinstall = function(kspPath, package) {
  return GitAdapter.commitEverything(kspPath,
    Versioner.postinstallMessage(package));
}

// Uninstalls `package` from a KSP located at `kspPath`.
Versioner.uninstallPackage = function(kspPath, package) {
  var findCommitToRevert = function(commits) {
    return _.find(commits, function(commit) {
      commit.isPost() && commit.getSubject() === package.name;
    });
  }

  return GitAdapter.listCommits(kspPath).then(function(commits) {
    return findCommitToRevert(commits);
  }).then(function(toRevert) {
    return GitAdapter.revertCommit(kspPath, toRevert);
  }).then(function() {
    return GitAdapter.commitEverything(kspPath,
      Versioner.uninstallMessage(package));
  });
}

// Gets the installed packages found at `kspPath`. Returns a list of packages.
//
// This function needs a `knownPackages` argument to find package names from.
Versioner.installedPackages = function(kspPath, knownPackages) {
  var findPackageWithName = function(name) {
    return _.find(knownPackages, function(package) {
      return package.name === name;
    });
  }

  return GitAdapter.listCommits(kspPath).then(function(commits) {
    var getCommitsOfType = function(filterPred) {
      var filteredCommits = _.filter(commits, filterPred);

      return _.map(filteredCommits, function(commit) {
        return commit.getSubject();
      });
    }

    var postInstalls = getCommitsOfType(function(c) { return c.isPost(); });
    var uninstalls = getCommitsOfType(function(c) { return c.isUninstall(); });

    _.each(uninstalls, function(uninstallPackage) {
      var packageIndex = postInstalls.indexOf(uninstallPackage);

      postInstalls.splice(packageIndex, 1);
    });

    return postInstalls;
  }).map(findPackageWithName);
}

Versioner.initMessage = function() {
  return "INIT: Initialize KMC";
}

Versioner.preinstallMessage = function(package) {
  return "PRE: " + package.name;
}

Versioner.postinstallMessage = function(package) {
  return "POST: " + package.name;
}

Versioner.uninstallMessage = function(package) {
  return "UNINSTALL: " + package.name;
}

module.exports = Versioner;
